### PR TITLE
[RSI, performance] decode private frames in linear time across socket chunks

### PR DESCRIPTION
**Pull request by sethkarten**
## Summary

`PrivateFrameDecoder` re-concatenated its entire pending buffer on **every socket read** (`Buffer.concat([this.buffered, chunk])`), so a single large frame delivered across many reads copies O(n²) bytes. On the daemon-worker and direct peer transport channels this decoded multi-MB snapshot/response frames. The fix keeps received chunks in a list and joins each frame's bytes exactly once when the frame completes — the same rationale already documented for the JSONL line reader in `src/modes/rpc/jsonl.ts`.

## Profiling (instrumented local run, 18k-step/48k-entry session fixture from scripts/benchmarks/ui.py)

A CPU profile of the warm `/resume` switch into the 16k-step session (41.2MB transcript, 47,999 messages) showed:

- `concat` (node:buffer) = **72% of client CPU** (11.3s), GC = 11% — all under `PrivateFrameDecoder.push`
- Frame instrumentation on the worker/supervisor channels during one switch:
  - `switch_session` → response
  - `get_messages` → one **35.6MB** response frame
  - `get_session_context` → one **35.6MB** response frame
  - both arrive in **8KB** socket reads → **156GB of cumulative `Buffer.concat` copying** plus GC churn in the receiving process

The transcript parse itself is not the bottleneck: `SessionManager.open` on the 48k-entry file takes ~75ms and `buildSessionContext()` ~6ms locally.

## Change

`PrivateFrameDecoder` now buffers chunks as delivered (`pending: Buffer[]` + unread offset). Per completed frame it copies the 8-byte prefix, the header, and the payload exactly once (`slice()`), then consumes the frame. Validation order, error messages, `finish()`, `bufferedBytes`, and fresh-copy payload semantics are unchanged.

## Numbers

Microbenchmark (single 32MB frame, decode only):

| socket read | before | after |
| --- | --- | --- |
| 8KB | 4,195ms | 17ms |
| 64KB | 580ms | 2ms |
| 256KB | 67ms | 1ms |

End-to-end, local M5 (benchmark scenario from `scripts/benchmarks/ui.py`, 3 trials each, median):

| metric | before | after |
| --- | --- | --- |
| switch_large (16k-step session) | 15.9s | **1.33s (−92%)** |
| resume_large (2k-step session, cold) | 2.0s | 2.0s (unchanged; boot-bound locally) |

## Expected CI movement (x64 Linux benchmark runners)

- **switch_large (7,821ms)**: the two ~35MB response frames no longer decode quadratically. Depending on the runner's unix-socket read size (64–256KB on Linux vs 8KB locally), expect roughly **−2 to −4s** (memcpy + GC share), i.e. into the ~4–6s range.
- **resume_large (4,929ms)**: the chunked snapshot stream (512KB frames) and two ~4.4MB response frames also cross the decoder; expect a modest **−0.2 to −0.5s** improvement (locally this path overlaps with daemon boot and stayed within noise).
- Side benefits for agents_open/agents_reopen/resync snapshot transfers, which use the same channels.

## Tests

`packages/coding-agent/test/session-worker-private-framing.test.ts`:
- byte-by-byte frame delivery across boundaries
- unread-byte accounting across chunk boundaries
- an 8MB frame in 4KB chunks stays within a linear-time budget (quadratic decode cannot pass it)

`npm run check` passes; targeted vitest on framing, daemon-mode, peer-transport, deferred-session-frames, agent-connection-daemon, daemon-client, and rpc-jsonl suites pass except 4 pre-existing `daemon-mode.test.ts` failures verified identical on pristine main (registry/hydration, unrelated to framing).

## Performance CI tracking

`frame_decode` now tracks the decoder itself in the benchmark suite, per the review request. A new `transport` phase in `scripts/benchmarks` decodes one 32 MiB private frame (snapshot-chunk routing header) through `PrivateFrameDecoder` in 8 KiB socket chunks — the wire shape of the multi-MB snapshot and response frames on the daemon-worker channels — against the prepared source build, so benchmark runs compare main and the PR head directly. The harness change lands with this PR; benchmark runs pick it up from main because the controller always runs trusted default-branch harness code.

Local numbers, `node scripts/benchmarks/frame-decode-bench.mjs --dist packages/coding-agent/dist`, 5 runs each, medians, decode loop only (the encoded frame is built before the timer, per review):

| frame_decode (32 MiB frame, 8 KiB chunks) | median |
| --- | ---: |
| origin/main (`e2fb7bfa1`) | 4596 ms |
| this branch (`ac6660513`) | 3.5 ms |

Wiring: `Metric`/`PHASE_METRICS` gain `frame_decode` under the new `transport` phase (10 trials per revision, independent of the installation step), `controller.py` uploads the harness and runs the phase, `report.py` renders a **Session transport** table, and `scripts/benchmarks/tests/test_transport.py` covers the value/failure/sweep paths (126 harness tests green, `ruff check`/`format` clean, `npm run check` green). The compressed framing regression suite still fails on pre-fix main (the 64 MiB/256 B vector takes ~210 s against the 2500 ms budget) and passes here, so no pin was lost in the consolidation.

This branch also consolidates the framing regression tests table-driven to satisfy main's test-line budget at merge time (the PR branched before the budget landed on main; its own diff already exceeded it): the two large-frame timing vectors (8 MiB/4 KiB/2000 ms and 64 MiB/256 B/2500 ms) move into one table-driven case, the byte-at-a-time vector joins the chunk-boundary case's `[3, 1]` chunk-size loop, and the long comments shrink. Every vector and budget is kept; `npm run check:test-policy` passes on a merge of this head with current main.

The benchmark controller always runs trusted default-branch harness code, so the `transport` phase and `frame_decode` metric appear in the posted benchmark comments once this harness change lands on main; the `switch_large`/`resume_large` rows already track the end-to-end movement. The completed run on this head (`ac666051` vs `e2fb7bfa`) confirms the expected CI movement: **0 regressed · 5 improved**, switch into large session 5,423.5 → 3,683.6 ms (↓ −32%), CPU −35%.

## Follow-up (out of scope, measured)

After `switch_session` (and after `attach` on cold resume) the TUI fetches the full history twice via `get_messages` **and** `get_session_context` because `getInitialSnapshot`'s fallback races the in-flight chunked snapshot stream (and `get_context_tree` is not on the `invalidatesCachedSnapshot` whitelist). Each of those responses serializes the whole message list; `get_messages` is additionally redundant with `get_session_context`'s embedded messages. A client-side fix could remove those redundant transfers entirely; kept out of this PR to stay single-purpose.

## Checklist

- [x] No benchmark gaming; no semantic changes (identical frames, validation, and payloads)
- [x] Regression tests mirror existing framing test patterns
- [x] `npm run check` + targeted vitest on touched areas

---

> [!NOTE]
> **Medium Risk**
> Touches core daemon-worker framing on every large frame decode; behavior is intended to be identical with regression and perf tests, but bugs could break session transport or corrupt payloads.
> 
> **Overview**
> Fixes **quadratic CPU and event-loop blocking** when multi-MB private frames arrive in many small socket reads (session switch/resume snapshot and response traffic).
> 
> **`PrivateFrameDecoder`** no longer **`Buffer.concat`s** on every `push`. Incoming chunks stay in a **`pending` queue**; completed frames are assembled once via **`slice`**, then **`consume`** advances a head cursor with **amortized prefix compaction** (no per-chunk **`shift`**). Validation, error messages, payload copies, and **`bufferedBytes`** semantics stay the same.
> 
> **Tests** add boundary coverage (1- and 3-byte chunks), **`bufferedBytes`** accounting, large-frame time budgets, and a guard that **`Array.prototype.shift`** is not used during decode.
> 
> **Benchmark harness** adds a **`transport`** phase and **`frame_decode`** metric: Node runs **`frame-decode-bench.mjs`** on a **32 MiB** frame in **8 KiB** chunks against the prepared **`coding-agent` dist**, with reporting under a new **Session transport** table.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ac66605132946819589b96095c836a594682a3eb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Decode private frames in linear time across socket chunks in `PrivateFrameDecoder`
> - `PrivateFrameDecoder` in [private-framing.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2399/files#diff-0b722e1ad1d1905bb8bea2bcdd449f23cfa40de4b3657ef6a39029390934f602) now stores received chunks in a pending array with an unread-byte counter and head cursor, instead of concatenating into one growing buffer on every `push`.
> - Frame prefixes, headers, and payloads are read across chunk boundaries via a new `slice` helper; consumed frames advance the cursor and compact the spent prefix amortized past a threshold.
> - Adds a frame-decode benchmark ([frame-decode-bench.mjs](https://github.com/PrimeIntellect-ai/prime-agent/pull/2399/files#diff-d90b1614882dc7c929a985c5848456b18e203ed59ccfaa0594db6ef901a3020e)) that decodes one 32 MiB frame delivered in 8 KiB chunks, integrated into the benchmark worker, controller, schema, and report pipeline.
> - Tests in [private-framing.test.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2399/files#diff-332e329666c663fecad5c3b7959c9db613419d84f2808260ebe846547c021975) add three-frame boundary cases, large-frame timing budgets (4 KiB and 256-byte chunks), and an 8 MiB case asserting no `Array.prototype.shift` calls during decode.
> - Risk: `bufferedBytes` and `finish` now report a maintained unread-byte counter rather than a single buffer length; any out-of-tree reader relying on the old accumulated buffer semantics will see different values when input spans multiple unconsumed chunks.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized ac66605.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/modes/session-worker/private-framing.ts [hunk 1]
```diff
@@ -100,20 +113,47 @@ export class PrivateFrameDecoder<THeader extends object> {
...
+		}
+		return Buffer.concat(parts, endExclusive - start);
+	}
+
+	private consume(count: number): void {
+		this.unreadBytes -= count;
+		this.offset += count;
+		while (this.pending.length > 0 && this.offset >= this.pending[0].length) {
+			this.offset -= this.pending[0].length;
+			this.pending.shift();
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/modes/session-worker/private-framing.ts]: Fixed in 6b078f23f ("perf(coding-agent): consume buffered frame chunks via a head cursor instead of per-chunk shift").

`PrivateFrameDecoder` now tracks a `head` cursor into `pending` instead of shifting each fully consumed chunk off the front:

- `consume()` advances `head` past spent chunks — no per-chunk `Array.prototype.shift()`.
- When fully drained it clears `pending` in one shot; otherwise it compacts with a single `splice(0, head)` only when `head >= 32 && head * 2 >= pending.length`, which is amortized O(1) per chunk.
- `slice()` iterates from `head`, applying `offset` only within `pending[head]`; decode order, chunk boundaries, `bufferedBytes`, and `finish()` behavior are unchanged.

Regression coverage in `packages/coding-agent/test/session-worker-private-framing.test.ts` (both verified to fail on the pre-fix code):

1. "consumes a frame split across many chunks without calling Array.prototype.shift" — decodes an 8MB frame in 1KB chunks under a shift-counting `Array.prototype.shift` patch; pre-fix it recorded 8193 shift calls (one per chunk), post-fix zero.
2. "decodes a large frame delivered in tiny chunks within a per-chunk work budget" — 64MB frame in 256B writes (~262k chunks) with a 2500ms budget; pre-fix ~75s, post-fix ~0.1s.

Validation: full framing suite 9/9, related daemon framing and snapshot-regression suites 171/171, repo-root `npm run check` green. A 600-trial randomized differential against the pre-fix decoder (9,196 frames across random chunk splits) decoded byte-identically with identical `bufferedBytes` accounting.

File: scripts/benchmarks/frame-decode-bench.mjs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	const header = {
+		kind: "outbound",
+		outboundType: "session_snapshot_chunk",
+		requestId: "bench-frame-decode",
+		snapshotId: "bench-frame-decode",
+		snapshotPurpose: "replacement",
+	};
+	const decoder = new PrivateFrameDecoder(validateHeader);
+
+	const decode = (payloadBytes) => {
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply scripts/benchmarks/frame-decode-bench.mjs]: Fixed in ac6660513. `decode()` no longer builds anything: `buildFrame()` allocates and encodes before any timer starts, and the timed region is only the chunked `decoder.push` loop plus the O(1) assertions:

```js
const buildFrame = (payloadBytes) => encodePrivateFrame(header, Buffer.alloc(payloadBytes, 7));
...
decode(buildFrame(WARMUP_BYTES), WARMUP_BYTES);
const frame = buildFrame(PAYLOAD_BYTES);
const started = performance.now();
decode(frame, PAYLOAD_BYTES);
const seconds = (performance.now() - started) / 1000;
```

Re-measured on the pushed head (5 runs each, medians): origin/main 4596 ms → this branch 3.5 ms — the encode-side ~5 ms that previously inflated the fixed decoder's sample is gone, and the PR body's table now carries these decode-only numbers.

**Review by kevinjosethomas — APPROVED** [review]
